src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs

```diff
@@ -95,6 +96,120 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
         public static FunctionPointerMethodSymbol CreateFromMetadata(CallingConvention callingConvention, ImmutableArray<ParamInfo<TypeSymbol>> retAndParamTypes)
             => new FunctionPointerMethodSymbol(callingConvention, retAndParamTypes);
 
+        public FunctionPointerMethodSymbol SubstiteParameterSymbols(TypeWithAnnotations substitutedReturnType, ImmutableArray<TypeWithAnnotations> substitutedParameterTypes)
+            => new FunctionPointerMethodSymbol(
+                this.CallingConvention,
+                this.RefKind,
+                substitutedReturnType,
+                this.RefCustomModifiers,
+                this.Parameters,
+                substitutedParameterTypes);
+
+        internal FunctionPointerMethodSymbol MergeEquivalentTypes(FunctionPointerMethodSymbol signature, VarianceKind variance)
+        {
+            var mergedParameterTypes = ImmutableArray<TypeWithAnnotations>.Empty;
+            bool hasParamChanges = false;
+            if (_parameters.Length > 0)
+            {
+                var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(_parameters.Length);
+                var paramVariance = variance switch
+                {
+                    VarianceKind.In => VarianceKind.Out,
+                    VarianceKind.Out => VarianceKind.In,
+                    _ => VarianceKind.None,
+                };
+
+                for (int i = 0; i < _parameters.Length; i++)
+                {
+                    var mergedParameterType = _parameters[i].TypeWithAnnotations.MergeEquivalentTypes(signature._parameters[i].TypeWithAnnotations, paramVariance);
+                    builder.Add(mergedParameterType);
+                    if (!mergedParameterType.IsSameAs(_parameters[i].TypeWithAnnotations))
+                    {
+                        hasParamChanges = true;
+                    }
+                }
+
+                mergedParameterTypes = builder.ToImmutableAndFree();
```

Comment:
>builder.ToImmutableAndFree() [](start = 39, length = 28)

Perhaps we should free the builder instead when there are no changes. We also often do not even allocate the builder until there is a change, but that is optional, I guess.